## <center>Тема “Обучение с учителем”

#### Задание 1

Импортируйте библиотеки pandas и numpy.

In [1]:
import warnings

warnings.filterwarnings('ignore')

In [2]:
import pandas as pd
import numpy as np

Загрузите "Boston House Prices dataset" из встроенных наборов данных библиотеки sklearn. Создайте датафреймы X и Y из этих данных.

In [12]:
from sklearn.datasets import load_boston
df=load_boston()
X=pd.DataFrame(df['data'], columns=df['feature_names'])
y=pd.DataFrame(df['target'], columns=["price"])

Разбейте эти датафреймы н а тренировочные (X_train, y_train) и тестовые (X_test, y_test) с помощью функции train_test_split так, чтобы размер тестовой выборки
составлял 30% от всех данных, при этом аргумент random_state должен быть равен 42.

In [16]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

Создайте модель линейной регрессии под названием lr с помощью класса LinearRegression из модуля sklearn.linear_model.

In [17]:
from sklearn.linear_model import LinearRegression
lr=LinearRegression()

Обучите модель на тренировочных данных (используйте все признаки) и сделайте предсказание на тестовых.

In [18]:
lr.fit(X_train, y_train)
pred=lr.predict(X_test)

Вычислите R2 полученных предказаний с помощью r2_score из модуля sklearn.metrics.

In [19]:
from sklearn.metrics import r2_score
R2 = r2_score(y_test, pred)
R2

0.711226005748496

#### Задание 2
Создайте модель под названием model с помощью RandomForestRegressor из модуля sklearn.ensemble.
Сделайте агрумент n_estimators равным 1000,
max_depth должен быть равен 12 и random_state сделайте равным 42.

In [20]:
from sklearn.ensemble import RandomForestRegressor
model = RandomForestRegressor(n_estimators=1000, max_depth=12, random_state=42)

Обучите модель на тренировочных данных аналогично тому, как вы обучали модель
 LinearRegression,
но при этом в метод fit вместо датафрейма y_train поставьте y_train.values[:, 0],
чтобы получить из датафрейма одномерный массив Numpy,
так как для класса RandomForestRegressor в данном методе для аргумента y предпочтительно применение массивов вместо датафрейма.

In [21]:
model.fit(X_train, y_train.values[:, 0])

RandomForestRegressor(max_depth=12, n_estimators=1000, random_state=42)

Сделайте предсказание на тестовых данных и посчитайте R2. Сравните с результатом из предыдущего задания.
Напишите в комментариях к коду, какая модель в данном случае работает лучше.

In [22]:
rf_pred=model.predict(X_test)
RF_R2=r2_score(y_test,rf_pred)
RF_R2

0.87472606157312

***RandomForestRegressor работает лучше чем LinearRegression. Качество модели RandomForestRegressor, оцененное с помощью метрики R2, составляет 0.87472606157312, в то время как линейная регрессия выдаёт 0.711226005748496*** 

#### *Задание 3
Вызовите документацию для класса RandomForestRegressor,
найдите информацию об атрибуте feature_importances_.

In [24]:
?RandomForestRegressor.feature_importances_

In [ ]:
С помощью этого атрибута найдите сумму всех показателей важности,
установите, какие два признака показывают наибольшую важность.

In [28]:
model.feature_importances_

array([0.03167574, 0.00154252, 0.00713813, 0.00123624, 0.01426897,
       0.40268179, 0.01429864, 0.06397257, 0.00528122, 0.01152493,
       0.01808108, 0.01245085, 0.41584732])

In [49]:
print(f'Сумма всех показателей важности признаков: {model.feature_importances_.sum()}')

Сумма всех показателей важности признаков: 1.0


In [50]:
feature_importances_top2=sorted(
    list(zip(df['feature_names'], model.feature_importances_)), key=lambda x: x[1])[-2:]

In [55]:
print(f'Наиболее важные ' 
      f'признаки\n{feature_importances_top2[1][0], feature_importances_top2[1][1]}'
      f'\n{feature_importances_top2[0][0], feature_importances_top2[0][1]}')

Наиболее важные признаки
('LSTAT', 0.4158473181914483)
('RM', 0.4026817857034993)
